Autopilot/Src/MPXV7002DP.cpp

```diff
@@ -0,0 +1,65 @@
+#include "airspeed.hpp"
+#include "adc.h"
+#include <stddef.h>
+#include <cmath>
+
+#define VS 5
+#define DENSITY_AIR 1.204
+#define ADC_RESOLUTION 4096
+
+
+MPXV7002DP::MPXV7002DP() {
+    //constructor
+    MX_ADC2_Init();
+}
+
+MPXV7002DP* MPXV7002DP::s_Instance = NULL;
+
+MPXV7002DP* MPXV7002DP::GetInstance() {
+    if(!s_Instance) {
+        s_Instance = new MPXV7002DP;
+    }
+    return s_Instance;
+}
+
+void MPXV7002DP::GetResult(airspeedData_t &Data) {
+        Data.airspeed = airspeed;
+        Data.isDataNew = dataNew;
+        dataNew = false;
+        Data.sensorStatus = sensorStatus;
+        Data.utcTime = timeOfLastResult;
+}
+
+void MPXV7002DP::Begin_Measuring() {
+        airspeed = readFromMPXV7002DP();
+        timeOfLastResult = getCurrentTime();
+        dataNew = true;
+        sensorStatus = 0;
+}
+
+void MPXV7002DP::Init() {
+
+    //do I need to init the adc?
+
+    uint16_t adcVal = HAL_ADC_GetValue(&hadc2);
+    offset = adcVal - (ADC_RESOLUTION-1)/2;
+}
+
+
+double MPXV7002DP::readFromMPXV7002DP() {
+    uint16_t adcVal = HAL_ADC_GetValue(&hadc2) - offset;
+
+    if(adcVal < ADC_RESOLUTION/2) {
+        return -std::sqrt( (-10000*(adcVal/(ADC_RESOLUTION-1) - 0.5))/DENSITY_AIR );
```

Comment:
I'm confused/there are some problems with the math here:
- What does 10000 mean ? I'd recommend using a define so people can understand what your'e doing.
- adcVal and ADC_RESOLUTION are both integers, so the compiler will do integer division and that division will be 0 every time. You can test this on some online cpp compiler. You'l have to cast one of the values to float or double to get the result you want.
- adcVal and ADC_resolution are always positive, right ? So won't the result of this square root always be NAN ?
- I don't understand why you need 2 seperate cases depending on the size of adcVal.